### ISSUE TITLE
Getting 2nd derivatives with respect to potential parameters

### ISSUE THREAD

**Issue by EqualAPriori** [openmm/openmm#2841]
Hi all,

This is more of a feasibility question for a new openmm feature.

I have been working on systematic coarse graining using the relative entropy, and one of the things we've found is that calculating second parameter derivatives, beyond its utility in numerical optimization, can actually help assess the statistical quality of coarse-graining ensembles (we to share the manuscript soon).  Because we're interested in getting the exact Hessian, typical numerical approximants used in optimization, like the Gauss-Newton matrix, are insufficient.

So far we have been calculating the Hessian manually, but because it essentially needs to do a lot of the calculations that are already handled in mature simulation code (e.g. looping over particles, handling of molecule topologies and forces), I am wondering if there might be some way to better leverage openMM's robust ecosystem of code. It is my understanding that OpenMM already has powerful internals for calculating certain global parameter first derivatives, even for custom forces. My question is how difficult would it be to extend this to obtain second derivatives? If it's not feasible, is there some way to expose part of what OpenMM calculates (e.g. pair distances, tracking of forcefield parameters) to another setting (probably python) that can leverage it and finish the post-processing? The goal is to minimize re-implementing things in a sub-optimal way.

If this is feasible I would love to help get this feature into either openMM or support software surrounding openMM. In addition to being useful in the coarse-graining work I'm doing, it seems like a general feature that should be very useful elsewhere as well -- e.g. to my understanding (and correct me if I'm wrong), ForceBalance, uses finite difference to estimate the second derivatives, but I'm not sure how well that would generalize to a coarse-graining setting that might have 100~1000 parameters (largely stemming from the fact that in coarse-grained models, mixing rules are less robust and having to resolve cross-interactions greatly increases the # of parameters to estimate).

Thanks for the help!
~Kevin

**Comment by peastman** [comment]
In principle it's certainly possible to compute second derivatives.  In most cases, the code would even be very similar to that for first derivatives.  You still loop over interactions, compute the terms you want for each one, and accumulate them in memory.  There are complications though.

For most standard bonded and nonbonded interactions it's not too complicated, but anything that involves a multistage computation becomes *very* complicated.  Examples of that would be GB implicit solvent forces and polarizable interactions.

Another issue is that you have to accumulate a lot more quantities.  If you want derivatives for N parameters, you have N first derivatives but N^2 second derivatives.  The current GPU kernels assume the number of parameters you want derivatives for is small enough that we can accumulate them in local variables.  That's fine for just a few parameters, but even 100 would cause it to run out of registers and dramatically hurt performance.  100^2 would be much worse, and would really need to be implemented in a different (and slower) way.

Also, we currently only support derivatives with respect to global parameters.  It sounds like you would want something different from that: perhaps per-particle parameters, or maybe the entries in a table of nonbonded parameters indexed by atom types?

**Comment by EqualAPriori** [comment]
Yes, we would need derivatives wrt per-particle parameters, or entries in a table of non-bonded parameters. One of the most important use-cases that we are interested in would be to see if derivatives can be calculated wrt atom/coarse-grained-bead charges. Another interesting use case is second derivatives involving collective variable potentials -- I imagine the efficiency of this will heavily depend on the intricacy of the collective variable.

Sounds like the bottleneck/limiting factor might be a memory issue -- is the problem that each GPU thread needs to track its own contribution to the N^2 second derivatives? Secondly, sounds like another important challenge is non-standard interactions, e.g. the GB and polarizable interactions -- either we have to restrict this 2nd derivative implementation to simpler interactions, or do the extra work of generalizing the computation.

Brainstorming some other considerations:
1. We wouldn't need this computation very frequently -- it'd either be used for post-processing, or if running alongside a simulation, definitely no more often than how frequently we write out trajectories. So perhaps slow won't be too bad (as long as it's not slower than serial processing.)
2. If calculating (and tracking) all the derivatives is too memory intensive to split onto GPU's, maybe it's be easier to implement a feature to save, e.g. the calculated pair distances and angles, which can then be post-processed later in another step/program outside of the typical pair-loop engine. What we do right now is to only save histograms (trading resolution in terms of # of bins for avoiding saving nparticle^2 values) of pair distances, angles, etc and post-process later. The reason why this can be effective is that the large number of parameters (e.g. the N>1000 total parameters that I cited) in coarse-grained models often derives from the large number of "parameters" in tabulated/spline potentials, rather than having lots of bead/atom types (typically ~10). It also allows for re-processing of data. This worked in serial processing, but I'm not sure if this would work in GPU processing (maybe still too memory intensive). Of course, if one further restricts the CG model to a simple functional form, e.g. Martini, then perhaps the low # of bead types means that a direct implementation akin to the first-derivative calculation in openMM would be sufficient, and that we needn't worry about N>100, so we should start with that.
3. Would some kind of post-processing step render a second-derivative calculation more feasible for the more complicated potential types, e.g. polarizable interactions? I'm not intimately familiar with the details of polarizable interactions, but it seems like at least drude oscillator implementations reduce to introducing a new bead type that could subsequently be treated as normal.
4. Perhaps to cut down on the N^2 cost for large N, it is inevitable that we have to enable an ability to to calculate only _certain_ cross-derivatives, e.g. decoupling bonded and non-bonded parameters.

Thanks again for the help!

**Comment by nbzy1995** [comment]
I am facing a similar situation as OP, and indeed I believe it is a common situation in research that we need to compute, for a  CustomNonbondForce, the second derivatives, and also pressure. 

My thoughts are:
- is it possible to exploit some automatic differentiation packages such as jax, to manipulate the potential function. To be specific, maybe I want to call `dforce=grad(force)` to obtain a new function, which can take the input coordinates and output a gradient vector of U(r) at this coordinates.
The time cost is not really a big problem because we won't call this during the simulation.

**Comment by peastman** [comment]
I think you're asking for a different feature?  The original request was second derivatives of the energy with respect to force field parameters.  It sounds like you're asking for second derivatives with respect to the particle coordinates?

**Comment by nbzy1995** [comment]
> I think you're asking for a different feature? The original request was second derivatives of the energy with respect to force field parameters. It sounds like you're asking for second derivatives with respect to the particle coordinates?

Oh yeh,  sorry. I will open a new issue.
